### PR TITLE
Add interpolation to player angle when moving around

### LINKED ISSUE #31: Add interpolation to the facing direction when moving
**I'm submitting a...**

- [ ] Bug report.
- [x] Feature request.

## Feature request ##

**Add interpolation to the player's angle instead of immediately snapping to the move direction**

I think the character controller is much more appealing when a bit of interpolation is added to the angle changing as the character moves around. Especially when using digital controls like a keyboard or dpad, the robotic looking snap as the character turns looks jarring.

I gave it a shot myself and recorded a [quick video](https://www.youtube.com/watch?v=BGg_KQH4ZjQ)

This was where I added the interpolation
Selection_381

### PR DESCRIPTION
**Pull request by FelixNemis**
**What is the current behavior?** 
Player immediately snaps to the angle of movement.

**What is the new behavior?**
Player's angle is interpolated between the current angle and facing in the direction of the movement. The interpolation factor is configurable and setting it to 0 restores the previous behaviour.

Close #31

### CODE REVIEW

File: src/Player/States/Move.gd [hunk 1]
```diff
@@ -9,6 +9,7 @@ export var max_speed: = 12.0
 export var move_speed: = 10.0
 export var gravity = -80.0
 export var jump_impulse = 25
+export(float, 0, 0.25, 0.005) var rotation_interpolation = 0.05
```
Comment (NathanLovato) [inline src/Player/States/Move.gd]: We use typed gdscript, please use `:=` instead of a single equal sign

File: src/Player/States/Move.gd [hunk 2]
```diff
@@ -33,7 +34,10 @@ func physics_process(delta: float) -> void:
 
 	# Rotation
 	if move_direction:
-		player.look_at(player.global_transform.origin + move_direction, Vector3.UP)
+		# To avoid dividing by 0, use 1 if rotation_interpolation equals 0
```
Comment (NathanLovato) [inline src/Player/States/Move.gd]: Better exclude 0 from the rotation interpolation. Possibly using a setter:
```gdscript
func set_angular_interpolation_rate(value: float) -> void:
    angular_interpolation = max(0.01, value)
```

File: src/Player/States/Move.gd [hunk 3]
```diff
@@ -33,7 +34,10 @@ func physics_process(delta: float) -> void:
...
+		var interpolation_factor: float = 1 if rotation_interpolation == 0 else delta / rotation_interpolation
```
Comment (NathanLovato) [inline src/Player/States/Move.gd]: This should be taken out

File: src/Player/States/Move.gd [hunk 4]
```diff
@@ -33,7 +34,10 @@ func physics_process(delta: float) -> void:
...
+		var target_direction: = player.transform.looking_at(player.global_transform.origin + move_direction, Vector3.UP)
+		player.transform = player.transform.interpolate_with(target_direction, interpolation_factor)
```
Comment (NathanLovato) [inline src/Player/States/Move.gd]: The interpolation factor should take time into account. Also, to ensure that the rotation happens at a fixed speed, you need to calculate it this way:
```gdscript
var lerp_rate := 1 - pow(angular_interpolation_rate, delta)
player.transform = player.transform.interpolate_with(target_direction, lerp_rate)
```

For a detailed explanation: https://www.construct.net/en/blogs/ashleys-blog-2/using-lerp-delta-time-924

Comment (FelixNemis) [reply src/Player/States/Move.gd]: The interpolation factor did factor in time but this is a better formula

**Comment by NathanLovato** [comment]
Actually my bad, I forgot that physics_process is framerate-dependent, it applies a fixed delta every frame so physics calculations stay stable. So the power doesn't do anything in this case, unlike in the process callback.

I'll fix and merge manually

**Comment by NathanLovato** [comment]
Merged manually as b059288deb960cec4b8e4979f1bd1cfec052bf9e
Thanks!

**Pull request closed without merging**